src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1898,6 +1911,119 @@ ErrorOnConcurrentRebalance(RebalanceOptions *options)
 }
 
 
+/*
+ * GetColocationId function returns the colocationId of the shard in a PlacementUpdateEvent.
+ */
+static int64
+GetColocationId(PlacementUpdateEvent *move)
+{
+	ShardInterval *shardInterval = LoadShardInterval(move->shardId);
+
+	CitusTableCacheEntry *citusTableCacheEntry = GetCitusTableCacheEntry(
+		shardInterval->relationId);
+
+	return citusTableCacheEntry->colocationId;
+}
+
+
+/*
+ * InitializeShardMoveDependencies function creates the hash maps that we use to track
+ * the latest moves so that subsequent moves with the same properties must take a dependency
+ * on them. There are two hash maps. One is for tracking the latest move scheduled in a
+ * given colocation group and the other one is for tracking the latest move which involves
+ * a given node either as its source node or its target node.
+ */
+static ShardMoveDependencies
+InitializeShardMoveDependencies()
+{
+	ShardMoveDependencies shardMoveDependencies;
+	shardMoveDependencies.colocationDependencies = CreateSimpleHashWithNameAndSize(int64,
+																				   ShardMoveDependencyInfo,
+																				   "colocationDependencyHashMap",
+																				   6);
+	shardMoveDependencies.nodeDependencies = CreateSimpleHashWithNameAndSize(int64,
+																			 ShardMoveDependencyInfo,
+																			 "nodeDependencyHashMap",
+																			 6);
+
+	return shardMoveDependencies;
+}
+
+
+/*
+ * GenerateTaskMoveDependencyList creates and returns a List of taskIds that
+ * the move must take a dependency on.
+ */
+static List *
+GenerateTaskMoveDependencyList(PlacementUpdateEvent *move, int64 colocationId,
+							   ShardMoveDependencies shardMoveDependencies)
+{
+	List *dependsList = NIL;
+
+	bool found;
+
+	/* Check if there exists a move in the same colocation group scheduled earlier. */
+	ShardMoveDependencyInfo *shardMoveDependencyInfo = hash_search(
+		shardMoveDependencies.colocationDependencies, &colocationId, HASH_ENTER, &found);
+
+	if (found)
+	{
+		dependsList = list_append_unique_ptr(dependsList,
+											 (void *) shardMoveDependencyInfo->taskId);
```

Comment:
Storing int64s in a List by pretending they are pointers seems a bit weird. I think using `CreateSimpleHashSet` is probably nicer for this usecase, given the numbers should also be unique.